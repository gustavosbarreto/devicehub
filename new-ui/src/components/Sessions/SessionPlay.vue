<template>
  <v-list-item
    @click="showDialog = !showDialog"
    v-bind="$attrs, $props"
    :disabled="notHasAuthorization"
  >
    <div class="d-flex align-center">
      <v-icon color="white" class="mr-2"> mdi-play </v-icon>

      <v-list-item-title data-test="mdi-information-list-item">
        Play Session
      </v-list-item-title>
    </div>
  </v-list-item>

  <v-dialog
    :transition="false"
    v-model="showDialog"
    max-width="1024px"
    min-width="350px"
  >
    <v-card class="bg-v-theme-surface">
      <v-card-title
        class="text-h5 pa-3 bg-primary d-flex justify-space-between align-center"
      >
        Watch Session
        <v-btn
          variant="text"
          data-test="close-btn"
          icon="mdi-close"
          @click="showDialog = false"
        />
      </v-card-title>

      <v-card-item class="ma-0 pa-0 w-100">
        <div ref="terminal" class="mt-n6 xterm-helper" />
      </v-card-item>

      <v-card-actions class="bg-v-theme-surface">
        <v-container>
          <v-row no-gutters>
            <v-col cols="2" sm="6" md="1">
              <div class="pt-4 ml-7">
                <v-icon
                  v-if="!paused"
                  variant="text"
                  icon="mdi-pause-circle"
                  class="pl-0"
                  color="primary"
                  rounded
                  size="x-large"
                  data-test="pause-icon"
                  @click="pauseHandler"
                />
                <v-icon
                  v-else
                  variant="text"
                  icon="mdi-play-circle"
                  class="pl-0"
                  color="primary"
                  rounded
                  size="x-large"
                  data-test="play-icon"
                  @click="pauseHandler"
                />
              </div>
            </v-col>

            <v-col cols="6" md="9">
              <div
                :elevation="0"
                class="pt-4 pl-9 mr-5 d-flex align-center"
                tile
              >
                <!-- <p class="mr-4">
                  {{ nowTimerDisplay }} - {{ endTimerDisplay }}
                </p> -->
                <v-slider
                  v-model="currentTime"
                  class="ml-0"
                  min="0"
                  :max="totalLength"
                  :label="`${nowTimerDisplay} - ${endTimerDisplay}`"
                  hide-details
                  color="primary"
                  data-test="time-slider"
                  @change="changeSliderTime"
                  @mousedown="(previousPause = paused), (paused = true)"
                  @mouseup="paused = previousPause"
                  @click="setSliderDiplayTime(currentTime)"
                />
              </div>
            </v-col>

            <v-col cols="6" md="2">
              <div :elevation="0">
                <v-select
                  :items="speedList"
                  v-model="defaultSpeed"
                  hide-details
                  prepend-icon="mdi-speedometer"
                  data-test="speed-select"
                  variant="underlined"
                  color="primary"
                  @change="speedChange(defaultSpeed)"
                />
              </div>
            </v-col>
          </v-row>
        </v-container>
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<script lang="ts">
import {
  computed,
  defineComponent,
  nextTick,
  onUpdated,
  ref,
  watch,
} from "vue";
import { Terminal } from "xterm";
import { FitAddon } from "xterm-addon-fit";
import moment from "moment";
import { useStore } from "../../store";
import { INotificationsError } from "../../interfaces/INotifications";

export default defineComponent({
  props: {
    uid: {
      type: String,
      required: true,
    },
    recorded: {
      type: Boolean,
      required: true,
    },
    notHasAuthorization: {
      type: Boolean,
      default: false,
    },
  },
  emits: ["update"],
  setup(props, ctx) {
    const showDialog = ref(false);
    const terminal = ref<any>(null);
    const currentTime = ref(0);
    const totalLength = ref(0);
    const endTimerDisplay = ref<string | number>(0);
    const getTimerNow = ref<string | number>(0);
    const paused = ref(false);
    const previousPause = ref(false);
    const sliderChange = ref(false);
    const speedList = ref([0.5, 1, 1.5, 2, 4]);
    const logs = ref([]);
    const frames = ref<any>([]);
    const defaultSpeed = ref(1);
    const transition = ref(false);
    const xterm = ref<any>();
    const fitAddon = ref<any>(null);
    const iterativeTimer = ref<any>();
    const iterativePrinting = ref<any>();

    const store = useStore();
    const length = computed(() => logs.value.length);
    const nowTimerDisplay = computed(() => getTimerNow.value);

    watch(showDialog, (value) => {
      if (!value) {
        close();
        showDialog.value = false;
      } else {
        displayDialog();
      }
    });

    onUpdated(() => {
      if (showDialog.value) {
        setSliderDiplayTime(currentTime.value);
      }
    });

    const openPlay = async () => {
      if (props.recorded) {
        await store.dispatch("sessions/getLogSession", props.uid);
        logs.value = store.getters["sessions/get"];
        // @ts-ignore
        totalLength.value = getSliderIntervalLength(null);
        setSliderDiplayTime(null);
        setSliderDiplayTime(currentTime.value);

        frames.value = createFrames();

        xterm.value = new Terminal({
          cursorBlink: true,
          fontFamily: "monospace",
        });

        fitAddon.value = new FitAddon();
        xterm.value.loadAddon(fitAddon.value); // adjust screen in container

        if (xterm.value.element) {
          xterm.value.reset();
        }
      }
    };

    const displayDialog = async () => {
      // await to change dialog for the connection
      try {
        await openPlay();

        await nextTick().then(() => {
          connect();
        });
      } catch {
        store.dispatch(
          "snackbar/showSnackbarErrorLoading",
          INotificationsError.sessionPlay
        );
      }
    };

    const connect = async () => {
      if (!xterm.value.element) {
        xterm.value.open(terminal.value);
        fitAddon.value.fit();
        xterm.value.focus();
        print(0, logs.value);
        timer();
      }
    };

    const getSliderIntervalLength = (timeMs: number | null) => {
      let interval;
      if (!timeMs && logs.value.length > 0) {
        // not params, will return metrics to max timelengtht
        // @ts-ignore
        const max = new Date(logs.value[length.value - 1].time);
        // @ts-ignore'
        const min = new Date(logs.value[0].time);
        // @ts-ignore
        interval = max - min;
      } else {
        // it will format to the time argument passed
        interval = timeMs;
      }

      return interval;
    };

    const setSliderDiplayTime = async (timeMs: number | null) => {
      const interval = getSliderIntervalLength(timeMs);
      const duration = moment.duration(interval, "milliseconds");

      // format according to how long
      let hoursFormat;
      if (duration.asHours() > 1) hoursFormat = "h";
      else hoursFormat = "";

      const displayTime = moment
        .utc(duration.asMilliseconds())
        .format(`${hoursFormat ? hoursFormat + ":" : ""}mm:ss`);
      if (timeMs) {
        endTimerDisplay.value = displayTime;
      } else {
        getTimerNow.value = displayTime;
      }
    };

    const createFrames = () => {
      // create cumulative frames for the exibition in slider
      let time = 0;
      let message = "";
      const arrFrames = [
        {
          // @ts-ignore
          incMessage: (message += logs.value[0].message),
          incTime: time,
        },
      ];

      for (let i = 1; i < logs.value.length; i += 1) {
        // @ts-ignore
        const future = new Date(logs.value[i].time);
        // @ts-ignore
        const now = new Date(logs.value[i - 1].time);
        const interval = moment
          // @ts-ignore
          .duration(future - now, "milliseconds")
          .asMilliseconds();
        time += interval;
        // @ts-ignore
        message += logs.value[i].message;
        arrFrames.push({
          incMessage: message,
          incTime: time,
        });
      }
      return arrFrames;
    };

    const speedChange = (speed: number) => {
      defaultSpeed.value = speed;
      xtermSyncFrame(currentTime.value);
    };

    const timer = () => {
      // Increments the slider
      if (!paused.value) {
        if (currentTime.value >= totalLength.value) {
          paused.value = true;
          return;
        }
        currentTime.value += 100;
        setSliderDiplayTime(currentTime.value);
      }
      iterativeTimer.value = setTimeout(
        timer.bind(null),
        100 * (1 / defaultSpeed.value)
      );
    };

    const changeSliderTime = () => {
      sliderChange.value = true;
      xtermSyncFrame(currentTime.value);
    };

    const pauseHandler = () => {
      paused.value = !paused.value;
      xtermSyncFrame(currentTime.value);
    };

    const close = async () => {
      transition.value = true;
      if (xterm.value) {
        xterm.value.reset();
        xterm.value.element?.remove();
      }
      clear();
      currentTime.value = 0;
      paused.value = false;
      defaultSpeed.value = 1;

      ctx.emit("update");
    };

    const clear = () => {
      // Ensure to clear functions for syncronism
      clearInterval(iterativePrinting.value);
      clearInterval(iterativeTimer.value);
    };

    const xtermSyncFrame = (givenTime: any) => {
      if (xterm.value) {
        xterm.value.write("\u001Bc"); // clean screen
        const frame = searchClosestFrame(givenTime, frames.value);
        clear();
        xterm.value.write(frame.message); // write frame on xterm
        iterativeTimer.value = setTimeout(timer.bind(null), 1);
        iterativePrinting.value = setTimeout(
          print.bind(null, frame.index + 1, logs.value),
          // @ts-ignore
          frame.waitForPrint * (1 / defaultSpeed.value)
        );
      }
    };

    const searchClosestFrame = (givenTime: any, frames: any) => {
      // applies a binary search to find nearest frame
      let between;
      let lowerBound = 0;
      let higherBound = frames.length - 1;
      let nextTimeSetPrint;

      for (; higherBound - lowerBound > 1; ) {
        // progressive increment search
        between = Math.floor((lowerBound + higherBound) / 2);
        if (frames[between].incTime < givenTime) {
          lowerBound = between;
          nextTimeSetPrint = givenTime - frames[between].incTime;
        } else {
          higherBound = between;
          nextTimeSetPrint = frames[between].incTime - givenTime;
        }
      }
      return {
        message: frames[lowerBound].incMessage,
        index: lowerBound,
        waitForPrint: nextTimeSetPrint,
      };
    };

    const print = (i: any, logsArray: any) => {
      // Writes iteratevely on xterm as time progresses
      sliderChange.value = false;
      if (!paused.value) {
        xterm.value.write(`${logsArray[i].message}`);
        if (i === logsArray.length - 1) return;
        const nowTimerDisplay = new Date(logsArray[i].time);
        const future = new Date(logsArray[i + 1].time);
        // @ts-ignore
        const interval = future - nowTimerDisplay;
        // @ts-ignore
        iterativePrinting.value = setTimeout(
          print.bind(null, i + 1, logsArray),
          interval * (1 / defaultSpeed.value)
        );
      }
    };

    return {
      showDialog,
      terminal,
      currentTime,
      totalLength,
      endTimerDisplay,
      getTimerNow,
      paused,
      previousPause,
      sliderChange,
      speedList,
      logs,
      frames,
      defaultSpeed,
      transition,
      pauseHandler,
      nowTimerDisplay,
      changeSliderTime,
      speedChange,
      setSliderDiplayTime,
    };
  },
});
</script>

<style lang="scss" scoped>
.xterm-helper {
  background: #0f1526;
  width: 105%;
}
</style>