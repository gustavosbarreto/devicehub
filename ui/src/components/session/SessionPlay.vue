<template>
  <fragment>
    <v-tooltip bottom>
      <template v-slot:activator="{ on }">
        <v-icon
          v-on="on"
          @click="openPlay()"
        >
          mdi-play-circle
        </v-icon>
      </template>
      <span>Play</span>
    </v-tooltip>
    <v-dialog
      v-model="dialog"
      max-width="1024px"
    >
      <v-card>
        <v-toolbar
          dark
          color="primary"
        >
          <v-toolbar-title>Play</v-toolbar-title>
          <v-spacer />
        </v-toolbar>
        <div ref="playterminal" />
        <v-card>
          <v-card-actions>
            <v-btn
              v-show="!disable"
              :disabled="disable"
              color="primary"
              class="mt-4"
              @click="connect()"
            >
              Play
            </v-btn>
            <v-slider
              v-model="currentTime"
              readonly
              min="0"
              :max="totalLength"
              :label="`${now} - ${end}`"
            />
          </v-card-actions>
        </v-card>
      </v-card>
    </v-dialog>
  </fragment>
</template>

<script>

import { Terminal } from 'xterm';
import { FitAddon } from 'xterm-addon-fit';
import moment from 'moment';
import 'xterm/css/xterm.css';

export default {
  name: 'SessionPlay',

  props: {
    uid: {
      type: String,
      required: true,
    },
  },

  data() {
    return {
      dialog: false,
      disable: false,
      currentTime: 0,
      totalLength: 0,
      timerDisplayEnd: null,
      timerDisplayNow: null,
    };
  },

  computed: {
    length() {
      return this.logs.length;
    },

    now() {
      return this.timerDisplayNow;
    },

    end() {
      return this.timerDisplayEnd;
    },

  },

  watch: {
    dialog(value) {
      if (!value) {
        this.close();
      }
    },
  },

  updated() {
    this.timerDisplayNow = this.duration(this.currentTime).display;
  },

  async created() {
    await this.$store.dispatch('sessions/getLogSession', this.uid);
    this.logs = this.$store.getters['sessions/getLogSession'];
    this.totalLength = this.duration(null).intervalLength;
    this.timerDisplayEnd = this.duration(null).display;
    this.timerDisplayNow = this.duration(this.currentTime).display;
    // eslint-disable-next-line no-console
    console.log(this.logs);
  },

  methods: {
    openPlay() {
      this.dialog = !this.dialog;
      this.xterm = new Terminal({ // instantiate
        cursorBlink: true,
        fontFamily: 'monospace',
      });
      this.fitAddon = new FitAddon(); // load fit
      this.xterm.loadAddon(this.fitAddon); // adjust screen in container
    },

    duration(timeMs) {
      let interval = 0;
      if (!timeMs) { // not params, will return metrics to max timelengtht
        const max = new Date(this.logs[this.length - 1].time);
        const min = new Date(this.logs[0].time);
        interval = max - min;
      } else { // it will format to the time argument passed
        interval = timeMs;
      }
      const duration = moment.duration(interval, 'milliseconds');
      const TimeObject = {
        seconds: Math.floor(duration.asSeconds()),
        minutes: Math.floor(duration.asMinutes()),
        hours: Math.floor(duration.asHours()),
      };
      return {
        display: `${TimeObject.hours}:${TimeObject.minutes}:${TimeObject.seconds}`, // format to slider label
        intervalLength: interval, // length of slider in Ms
      };
    },

    timer() { // Increments the slider
      if (this.currentTime >= this.totalLength) return;
      this.currentTime += 100;
      this.iterativeTimer = setTimeout(this.timer.bind(null), 100);
    },

    connect() {
      this.disable = true;
      this.xterm.open(this.$refs.playterminal);
      this.$nextTick(() => this.fitAddon.fit());
      this.fitAddon.fit();
      this.xterm.focus();
      this.print(0, this.logs);
      this.timer();
      if (this.xterm.element) { // check already existence
        this.xterm.reset();
      }
    },

    close() {
      if (this.xterm) this.xterm.dispose();
      this.disable = false;
      clearInterval(this.iterativePrinting);
      clearInterval(this.iterativeTimer);
      this.currentTime = 0;
    },

    print(i, obj) {
      this.check_hostname(i);
      if (i === this.logs.length - 1) return;
      const nextObj = this.logs[i + 1];
      const now = new Date(obj.time);
      const future = new Date(nextObj.time);
      setTimeout(this.print.bind(null, i + 1, nextObj), future.getTime() - now.getTime());
    },

    check_hostname(i) {
      if (this.logs[i].message.includes('@')) { // hostname verify,
        this.xterm.write(`${this.logs[i].message}`);
      } else {
        this.xterm.write(`${this.logs[i].message}\r\n`);
      }
    },
  },
};

</script>
