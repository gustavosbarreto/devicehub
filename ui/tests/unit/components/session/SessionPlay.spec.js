import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import SessionPlay from '@/components/session/SessionPlay';

describe('SessionPlay', () => {
  const localVue = createLocalVue();
  localVue.use(Vuex);

  let wrapper;

  const isOwner = true;
  const uid = '8c354a00f50';
  const recorded = true;
  const speedList = [0.5, 1, 1.5, 2, 4];

  const session = [
    {
      uid: '1a0536ab37d',
      message: '\u001b]0;shellhub@shellhub: ~\u0007shellhub@shellhub:~$ ',
      tenant_id: 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx',
      time: '2020-09-24T18:32:04.559Z',
      width: 110,
      height: 23,
    },
    {
      uid: '1a0536ab37d',
      message: 'l',
      tenant_id: 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx',
      time: '2020-09-24T18:32:06.181Z',
      width: 110,
      height: 23,
    },
    {
      uid: '1a0536ab37d',
      message: 's',
      tenant_id: 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx',
      time: '2020-09-24T18:32:06.344Z',
      width: 110,
      height: 23,
    },
    {
      uid: '1a0536ab37d',
      message: '\r\n',
      tenant_id: 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx',
      time: '2020-09-24T18:32:06.592Z',
      width: 110,
      height: 23,
    },
    {
      uid: '1a0536ab37d',
      message: '\u001b]0;shellhub@shellhub: ~\u0007shellhub@shellhub:~$ ',
      tenant_id: 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx',
      time: '2020-09-24T18:32:06.61Z',
      width: 110,
      height: 23,
    },
  ];

  const store = new Vuex.Store({
    namespaced: true,
    state: {
      isOwner,
      session,
    },
    getters: {
      'namespaces/owner': (state) => state.isOwner,
      'sessions/get': (state) => state.session,
    },
    actions: {
      'sessions/getLogSession': () => {
      },
    },
  });

  beforeEach(() => {
    wrapper = shallowMount(SessionPlay, {
      store,
      localVue,
      stubs: ['fragment'],
      propsData: { uid, recorded },
      mocks: {
        $env: (isEnterprise) => isEnterprise,
      },
    });
  });

  it('Is a Vue instance', () => {
    expect(wrapper).toBeTruthy();
  });
  it('Renders the component', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('Receive data in props', () => {
    expect(wrapper.vm.uid).toEqual(uid);
    expect(wrapper.vm.recorded).toEqual(recorded);
  });
  it('Process data in the computed', () => {
    expect(wrapper.vm.length).toEqual(0);
    expect(wrapper.vm.nowTimerDisplay).toEqual(0);
  });
  it('Compare data with default value', () => {
    expect(wrapper.vm.dialog).toEqual(false);
    expect(wrapper.vm.currentTime).toEqual(0);
    expect(wrapper.vm.totalLength).toEqual(0);
    expect(wrapper.vm.endTimerDisplay).toEqual(0);
    expect(wrapper.vm.getTimerNow).toEqual(0);
    expect(wrapper.vm.paused).toEqual(false);
    expect(wrapper.vm.previousPause).toEqual(false);
    expect(wrapper.vm.sliderChange).toEqual(false);
    expect(wrapper.vm.speedList).toEqual(speedList);
    expect(wrapper.vm.logs).toEqual([]);
    expect(wrapper.vm.frames).toEqual([]);
    expect(wrapper.vm.defaultSpeed).toEqual(1);
    expect(wrapper.vm.transition).toEqual(false);
  });
});
