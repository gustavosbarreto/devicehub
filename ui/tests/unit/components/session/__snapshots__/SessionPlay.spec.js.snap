// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SessionPlay Dialog opened Renders the component 1`] = `
<fragment-stub data-test="render-fragment"><span><button type="button" class="v-icon notranslate v-icon--link mdi mdi-play-circle theme--light"></button></span><span class="v-tooltip v-tooltip--bottom"><!----></span>
  <div role="dialog" class="v-dialog__container"></div>
</fragment-stub>
`;

exports[`SessionPlay Dialog opened is not owner Renders the component 1`] = `""`;

exports[`SessionPlay Icon Renders the component 1`] = `
<fragment-stub data-test="render-fragment"><span><button type="button" class="v-icon notranslate v-icon--link mdi mdi-play-circle theme--light"></button></span><span class="v-tooltip v-tooltip--bottom"><!----></span>
  <div role="dialog" class="v-dialog__container">
    <!---->
  </div>
</fragment-stub>
`;
